### ISSUE TITLE
future<future<>> unwrapping

### ISSUE THREAD

**Issue by brycelelbach** [executors/futures#29]
The Concurrency TS v1 futures do some unwrapping of future<future<>>s for convenience. We need to think about addressing this as it is pretty fundamental to usability.

We probably need to bug @K-ballo about this.

**Comment by griwes** [comment]
Are we discussing this in general, or just for constructors, or...?

Bryce knows my thoughts about the decision EWG made on CTAD, and I think we could possibly go in the opposite direction for the sake of the feature, and have a ctor that deduces actually wrap - but I can definitely be convinced that the cost of being inconsistent is too high.

If we do any sort of unwrapping, and we definitely should, we should also *expose* a library facility to flatten futures; I call that `join`, but I'm okay with `unwrap` or `flatten` or whatever. 

IMHO `.then` **should** unwrap, because that's what I want in 99% of cases; OTOH there are cases where I explicitly don't want that, and I have a mode called something like `do_not_unwrap` that... doesn't unwrap (ask David Sankel why this is truly necessary; basically without this mode you can't get a `Functor` out of a future, and I think that that is bad):

```
future.then(foo); // unwraps
future.then(do_not_unwrap, foo); // doesn't unwrap
```

I'd also be okay with a separate function name for the non-unwrapping behavior. I'd leave `.then` for the unwrapping one, since that is the good name, and I believe unwrapping to be the common case.

**Comment by LeeHowes** [comment]
Continuations should unwrap on return. Any returned Future/SemiFuture from a continuation is effectively returnedFuture.via(executor then is executing on) and such that the converted returned future will satisfy the next future in the main continuation chain. Without that we can't really write async continuations, because it's very different behaviour from returning a wrapped future that we have to explicitly call a continuation on later.

I'm not sure about the use cases for generally unwrapping a future of a future, except that if you allow a future of a future then that future-returning continuation is ambiguous.

**Comment by dhollman** [comment]
>  OTOH there are cases where I explicitly don't want that

My comment on #8 addresses another use case for this: child-to-child dependencies using `.then` for chaining. That said, I think I agree that the unwrapping behavior is what you want in most cases and should be the default.

**Comment by LeeHowes** [comment]
In talking to people here today, I think Michal has a point. There is a strong argument for having separate functor and monad forms of then. What is less clear to me, is whether we want to allow a value return from the monad form, or whether we should require that people use the functor form if they want to do that. That would really mean we were explicit about calling the unwrapping version, rather than the reverse.

future.bind([](){return int;}); // invalid
future.then(do_not_unwrap, foo); // doesn't unwrap

**Comment by griwes** [comment]
I don't think we need `bind` to be a part of the core interface; it is sufficient to have `.then(do_not_unwrap)` and `join` to implement `bind` outside of `future`. (And `join` can be implemented separately.) The unwrapping `.then` is really just a convenience "do what I mean" feature.

And when you write `bind` as `join(fut.then(do_not_unwrap, ...))`, it is obvious that it's going to fail when there's nothing to unwrap.

I agree that the notation I'm proposing isn't ideal, but if you actually want to directly use the functor and monad interface, you're probably going to have your own wrappers that verify if the world is sane.

**Comment by LeeHowes** [comment]
I think that can work.

**Comment by RedBeard0531** [comment]
Just to be clear, this is about adding methods, not removing the current behavior of then() to produce the same return type for a callback that returns U or future<U>, right?

Do we have any real use case for then() returning `future<future<T>>` other than "it would be nice if future was a Functor"? That type is weird for the much the same reason that `future<std::expected<T>>` and `future<std::exception_ptr>` are (we [ban](https://github.com/RedBeard0531/mongo_utils/blob/5d24c3dce16d99783255879fb72d8442003d57b8/util/future.h#L622-L625) making a future with any of our equivalents). Having multiple layers of errors or waiting is both awkward to work with and increases the risk of confusion and bugs. It is generally unclear how those types are semantically different from `future<T>` /`future<void>`. I'm hesitant to start offering much official blessing of those types and I think we should make sure that our interfaces never require them to come into existence.

Given that you could always make your own `map()` by using an explicit `promise<future<T>` that you complete from a callback, I think I'm -1 on this idea until we have a real use case.

**Comment by griwes** [comment]
We are talking about adding an overload.

Forcing a user who just wants to have the API of a functor to use the unwieldy nonsense that is `std::promise` IMO goes against everything we are trying to do here. IMHO users should have to almost never actually use `std::promise` - I see it as a very low-level building block.

Besides, as far as my experience tells me, that overload is already present, and is how you implement the "normal" `.then` - at least that's how I do it (do not unwrap, and then join when you can).

**Comment by RedBeard0531** [comment]
We implement the two overloads of then [completely independently](https://github.com/RedBeard0531/mongo_utils/blob/5d24c3dce16d99783255879fb72d8442003d57b8/util/future.h#L768-L827). This both ensures the best performance and avoids creating `Future<Future<T>>` which we ban. I used a similar implementation in my [single threaded archetype PR](https://github.com/RedBeard0531/futures-impl/blob/0bc228bdb565b19d2bb3a63fdb8f31c22420f497/archetypes/single_threaded_executor.cpp#L126-L162), but since it is in c++17 I used `if constexpr` rather than two overloads.

Can you explain your use case for `Future<Future<T>>` where always implicitly collapsing to `Future<T>` is undesirable? In the absence of a real use case, I think the type is problematic enough that I consider it a *good* thing if users need to go out of their way to get one.

That said, I think it is possible to offer a less painful promise api than `std::promise` and I think that improving that API is worth doing. While I agree that promises should be used much less often then futures, I think they will still be used frequently enough to make them good. For example, it is hard to combine multiple futures where the first one to finish wins, without using a promise.

**Comment by griwes** [comment]
I don't have a concrete use case.

*However*, as an avid metaprogrammer, I want to scream in pain when someone forbids a specialization of a type that otherwise meets all the requirements (in this case, `Movable`). And also from a generic programming perspective, having no way to say whether you want the call to unwrap or not, is going to lead us towards some incredibly hard to analyze situations inside generic algorithms.

Which may be *fine* in your specific use case, but for the general public, I have a very strong objection to not giving the user full control, *for the sake of generic code*. I know I'm not the one. (And there's even people who think that an API that is neither (or both, depending on how you look at it) Functor.map nor Monad.bind is just wrong anyway...).

You can call this an ideological argument, and it might well be, but having seen us (= the committee) do terrible things to symmetry in generic code, I really don't want us to repeat anything like that. If we have an API that does magic - that's fine. That's often what's necessary, for the sake of the end user. But I believe we shouldn't sacrifice the middle user for the sake of the end user - which is why I want us to have a toy for the middle user, too.

The implementation you have for that is interesting. I will need some more time to understand if and how it works with cases that I deeply care about (that is, shared futures) and I'll definitely try to see whether I can do something similar in my implementation. But that doesn't change my argument, and is merely an optimization, which doesn't change the fact that `.then(do_not_unwrap)`, or however you spell it, is the base privimite that you can build all others from. (Which is something that I always insinctively felt, but David Sankel has put it together nicely in his talk at C++Now last year: https://www.youtube.com/watch?v=2OY0Zn3oBCE.)

(Sorry if this feels a little disorderly, it's late where I am and this is pretty much a brain dump.)

**Comment by dhollman** [comment]
> Can you explain your use case for `Future<Future<T>>` where always implicitly collapsing to `Future<T>` is undesirable? In the absence of a real use case, I think the type is problematic enough that I consider it a good thing if users need to go out of their way to get one.

Okay, I'll give it a shot. Suppose you have a generic method that strings two callbacks together, much in the same way `.then()` itself does, but with some special semantics on the relationship between the two callables (say, like running them on the same thread or hardware resource or something).  One implementation of this could look something like:

```c++
template <typename Executor, typename CallbackA, typename CallbackB>
auto special_string_together(Executor ex, CallbackA&& a, CallbackB&& b) {
  auto ex_sp = execution::require(ex, special_semantics);
  return ex_sp.twoway_execute(std::forward<CallbackA>(a)).then(
    [b=std::move(b)](auto a_val){
      std::forward<CallbackB>(b)();
      return a_val;
    }
  );
};
```

That works fine in most cases. But then what happens when someone writes this code:

```c++
special_string_together(ex,
  [ex]{
    cout << "hello" << endl;
    auto f1 = ex.twoway_execute([]{ sleep(30s); });
    cout << "goodbye" << endl;
    return std::move(f1);
  },
  [ex]{
    cout << "world" << endl;
    auto f2 = ex.twoway_execute([]{ sleep(30s); });
    cout << "moon" << endl;
  }
);
```

(Replace the output with some feedback to the user and the sleeps with some operation of unknown length and you have a real use case).

On machines with terminal output rates faster than one character every 5 seconds, you'd usually expect something like:

```
hello
world
goodbye
moon
```

and you'd expect most efficient executions to take about 30 seconds.

But that's not what you get if you *always* have automatic unwrapping and forbid `future<future<T>>`. Instead, the program takes 60 seconds to run, and you get

```
hello
goodbye
world
moon
```

In other words, your generic function really did (for some specific reason) want to string together the parent tasks, but instead the behavior you get is connecting the first child to the second parent. This isn't necessarily an incorrect execution of the program, but it is *performance incorrect* due to semantics mismatch, which is almost the same thing.

Now, why wouldn't you just write your program with the inner tasks in a different place? Maybe they're private implementation details of two different APIs that don't know about each other. (Putting the callbacks inline like this makes it look pretty silly, but it doesn't always work that way. Encapsulation structure of a program and dependency structure of a program don't always line up perfectly).

I have not listened to David Sankel's talk on the matter yet, but I plan to try and listen to it on my bike ride home today. It's entirely possible he's saying the same thing, in which case this was unnecessary.

**Comment by LeeHowes** [comment]
When unwrapping a future, where should the unwrapper run?

It strikes me that enqueuing an unwrapping operation on an executor with a weak agent is a problem - the unwrapping operation is fundamentally an enqueue in itself isn't it? Something like:

```
Future<int> unwrap(Future<Future<int>> f) {
  std::standard_promise<int> p;
  auto newF = p.get_semi_future();
  auto e = f.get_executor();
  f.then(
    [e, p = std::move(p)](Future<int> f2){
      f2.then([p = std::move(p)](int val){
        p.set_value(val);
      }
    });
  return newF;
}
```

such that the futures returned by both .then calls are discarded with detach behaviour (an internal version of .then could of course do that too using a oneway executor). 

Enqueuing that continuation to do the unpacking has to be a lock-free operation or the calling context has to have strong forward progress guarantees. That means that the entire unwrap operation has to run on an executor with strong forward progress guarantees or e has to be lock-free.

**Comment by griwes** [comment]
`unwrap` should have an overload taking an executor specifying where to enqueue that, in case the user knows he needs a stronger executor in order for this operation to succeed.

**Comment by brycelelbach** [comment]
Does `hpx_future<cuda_future<int>>` unwrap? E.g. do heterogeneous future types unwrap?

**Comment by LeeHowes** [comment]
If unwrap uses continuations, and if unwrap takes an executor, then I think any heterogeneous future can unwrap.

If not, then only some futures can unwrap, because not all futures make sense to do any other form of unwrapping.

For example, if I have a Future<Future<T>> and the inner future simply does not exist, with no shared state or other data, until it is received by the continuation attached to the outer future, I really can't do any magic unwrapping.

**Comment by brycelelbach** [comment]
At the NVIDIA F2F meeting, we came up with the idea of having future wrapper helpers (see issue #48) to help us build low-level executor `FutureContinuation`s. As part of that discussion, we felt that we could use a similar solution for future unwrapping. Then, we took this poll:

Poll: `future<future<>>` unwrapping is a version 1 feature.
SF F N A SA
1 1 2 0 3

We agreed that for version 1, we should not do unwrapping; unwrapping can be achieved with a future continuation helper function.

However, on the 05-01-18 phone call, we were unclear if the above strategy works.

We need to have a plan at least for version 1 - if the plan is to punt, that's fine, but we need to be confident how we'll punt and how we'll do future unwrapping later.

@RedBeard0531 also suggested a more radical option of spelling `.then` differently in version 1, with the assumption we would add a more user-friendly unwrapping `.then` later. @brycelelbach is concerned that requiring future authors to do unwrapping may be a burden.

Action Items
* [ ] We need to have a plan for version 1 (that plan could be to punt though).

**Comment by griwes** [comment]
My concern here is that some of those things are going to have the same implementations everywhere; we need to think about how to do some form of mixins for implementers of future types, so that they don't have to keep reimplementing all this.

**Comment by RedBeard0531** [comment]
> We could define the concept in terms of map and bind and provide a do-what-I-mean wrapper, but I don't know how to do that in a sensible manner when it comes to the syntax the end user would have to use.

The "nuclear option" is that the *concept* won't provide *any* do-what-I-mean wrapper. It delegates that portion of API design to the concrete future implementations which may be in a better place to address the needs of their end users. 

>My concern here is that some of those things are going to have the same implementations everywhere; we need to think about how to do some form of mixins for implementers of future types, so that they don't have to keep reimplementing all this.

I assume there will be several orders of magnitude more code using futures than implementing them, so I don't think it really makes sense to optimize for ease of implementation, or to standardize any mixins to make it easier to implement. If any organisation writes enough futures for that to be worth it, they are free to do so as an implementation strategy. They could even open source it as a third-party lib.

**Comment by jaredhoberock** [comment]
> My concern here is that some of those things are going to have the same implementations everywhere; we need to think about how to do some form of mixins for implementers of future types, so that they don't have to keep reimplementing all this.

In order to avoid forcing every future implementer to redundantly have to implement e.g. `.then()`:

Have you all considered making something like `std::then(future, f)` a free function or customization point?

The future type returned by `std::then` could be a type that is guaranteed to always have a `.then()` member function. It could do this by adapting the future type returned by `execution::then_execute(ex, future, f)` or whatever primitive is used to actually implement `then`.

**Comment by griwes** [comment]
I'm starting to be additionally sympathetic to the suggestion from @LeeHowes that we use an operator. We *could* then have a sufficiently-global `operator|(ContinuableFuture, FutureContinuation)` that'd solve the problem of having everybody reimplement the world.

We'd need to have the same lengthy discussion again, this time about "how do we embed the information controlling unwrapping in this", but I think that's doable, thanks to the helper functions we want to have.

**Comment by griwes** [comment]
Alternatively, this discussion proves that we should kill two birds with one stone:
1. Not have a "do what I mean" for now means clear semantics on the functions that we'd propose (I'd be sad, though), and
2. Address @hkaiser 's concerns from #57.

Although I am still of the mind that we should have just a single function, and control its behavior through concrete `FutureContinuation` types. I don't believe we can design this well for V1, though.

**Comment by hkaiser** [comment]
> Does hpx::future<cuda_future<int>> unwrap? E.g. do heterogeneous future types unwrap?

@brycelelbach Yes. that unwraps as we have made sure the futures returned from CUDA are compatible.

**Comment by hkaiser** [comment]
FWIW, unwrapping `future<future<>>` (and `future<shared_future<>>` for that matter) through an unwrapping constructor works very nicely for us (in HPX). We have not encountered any problems with this so far.

**Comment by LeeHowes** [comment]
> Although I am still of the mind that we should have just a single function, and control its behavior through concrete FutureContinuation types.

We should, if only because it allows easier free functions/operator use. The options then should be:
 * Never unwrap
 * Always unwrap

One question for me is how we define mutually unwrappable types. Maybe the cleanest thing there is to depend on an explicit unwrapping constructor on the future and check for its existence, which would align with what Hartmut suggested. Obviously in that case we can have people use the constructor explicitly, but if we have:
` get_future().then(unwrap(then_value(...))).then(...)`
or:
` get_future().then(then_value_unwrap(...)).then(...)`

unwrap would cause then to implicitly produce:
`AFuture<AnotherFuture<type>>`

and would internally call:
`AFuture<type>{AFuture<AnotherFuture<type>>)`

if that does not compile, then unwrapping is rejected statically.

` get_future().then(then_value(...)).then(...)`

would always return `AFuture<AnotherFuture<type>>`

Does that align with others' thoughts? You can always produce a future that unwraps nothing, not even itself.

**Comment by LeeHowes** [comment]
We could further:
 * Require that all futures be able to unwrap themselves. Future<Future<T>> -> Future<T> That gives some basic functionality, and clearly an implementer of a future knows how to most efficiently do that.
 * We could require that we can use the unwrap operation, as above, for any heterogeous future pairing. That means we can always have a future optimise by having multiple unwrapping constructors, but in the worst case it would mean chaining the inner future onto the outer executor to get the right future type.

The latter might make performance predictability hard.

**Comment by brycelelbach** [comment]
Moving to V1.5 (based on the poll from the 04-26-18 NVIDIA F2F meeting).
